diff -up rpm-4.8.0/rpmrc.in.em64t rpm-4.8.0/rpmrc.in
--- rpm-4.8.0/rpmrc.in.em64t	2010-05-04 10:50:37.000000000 +0300
+++ rpm-4.8.0/rpmrc.in	2010-05-04 10:51:28.000000000 +0300
@@ -96,6 +96,7 @@ arch_canon:	i386:	i386	1
 arch_canon:	x86_64:	x86_64	1
 arch_canon:	amd64:	amd64	1
 arch_canon:	ia32e:	ia32e	1
+arch_canon:	em64t:	em64t	1
 
 arch_canon:	alpha:	alpha	2
 arch_canon:	alphaev5: alphaev5	2
@@ -351,9 +352,9 @@ arch_compat: s390x: s390 noarch
 
 arch_compat: ia64: noarch
 
-arch_compat: x86_64: amd64 athlon noarch
-arch_compat: amd64: x86_64 athlon noarch
-arch_compat: ia32e: x86_64 athlon noarch
+arch_compat: x86_64: amd64 em64t athlon noarch
+arch_compat: amd64: x86_64 em64t athlon noarch
+arch_compat: ia32e: x86_64 em64t athlon noarch
 
 arch_compat: sh3: noarch
 arch_compat: sh4: noarch
